lib/mixduty/response.ex

```diff
@@ -0,0 +1,146 @@
+defmodule Mixduty.Response do
+  require Logger
+  alias Mixduty.Error
+
+  alias Jason, as: JSON
+
+  @enforce_keys [:body, :headers, :status_code, :request, :request_url]
+
+  defstruct [
+    :body,
+    :headers,
+    :status_code,
+    :request,
+    :request_url
+  ]
+
+  @type headers ::
+          [{atom(), binary()}]
+          | [{binary(), binary()}]
+          | %{required(binary()) => binary()}
+          | any()
+
+  @type t :: %__MODULE__{
+          body: Map.t() | String.t(),
+          headers: headers(),
+          status_code: integer(),
+          request: HTTPoison.Request.t(),
+          request_url: binary() | any()
+        }
+
+  def new({:ok, %HTTPoison.Response{status_code: code, body: body} = resp})
+      when code in 200..299 and body in [nil, ""] do
+    resp
+    |> Map.from_struct()
+    |> Map.put(:body, %{})
+    |> put_into(__MODULE__)
+  end
+
+  def new({:ok, %HTTPoison.Response{status_code: code, body: body} = resp})
+      when code in 200..299 do
+    case json_decode(body) do
+      {:ok, decoded_body} ->
+        resp
+        |> Map.from_struct()
+        |> Map.put(:body, decoded_body)
+        |> put_into(__MODULE__)
+
+      {:error, error} ->
+        {:error, %Error{message: "JSON parse error: #{error}", status_code: code, cause: resp}}
+    end
+  end
+
+  def new({:ok, %HTTPoison.Response{headers: headers, status_code: code} = resp})
+      when code in 400..599 do
+    {:error,
+     %Error{
+       message: reason_from_status_header_or_code(headers, code),
+       status_code: code,
+       cause: resp
+     }}
+  end
+
+  def new({:ok, %HTTPoison.Response{status_code: code} = resp}) do
+    {:error, %Error{message: "Unhandled status code: #{code}", status_code: code, cause: resp}}
+  end
+
+  def new({:error, %HTTPoison.Error{reason: reason} = error}) do
+    {:error, %Error{message: "HTTP request failed. Reason: #{inspect(reason)}", cause: error}}
+  end
+
+  def put_into(response, struct_name_or_map, opts \\ [])
+
+  def put_into({:ok, %__MODULE__{body: body}}, %{} = map, opts),
```

Comment:
Are you sure that first parameter is %{:ok, something}, Map.put returns a map https://github.com/PagerDuty/mixduty/pull/21/files#diff-c156091a837a956795e7aa94c3a8983de1f1ad3c14c18c74f2ab279e01340f83R35

Reply:
We have this matching for this case
```
def put_into(%{} = map, struct_name, _opts), do: {:ok, struct(struct_name, map)}
```